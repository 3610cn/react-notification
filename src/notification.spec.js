import React, { createElement } from 'react';
import Notification from './notification';
import mockNotification from '../test/mockNotification';

describe(Notification.name, () => {
  // The component class names, used to find elements
  const actionClassName = '.notification-bar-action';
  // const messageClassName = '.notification-bar-message';
  // const titleClassName = '.notification-bar-title';

  /**
   * TEST 1a: message property with type string
   * This test ensures the most basic mounted component matches the
   * stored snapshot.
   */
  it('should render component with message type string', () => {
    expect(mount(
      <Notification
        message={mockNotification.message}
      />
    )).toMatchSnapshot();
  });

  /**
   * TEST 1b: message property with type element
   * The `message` property should accept a React element.
   */
  it('should render component with message type element', () => {
    expect(mount(
      <Notification
        message={createElement('div', {
          className: 'notification-bar-custom-message'
        }, 'Foo')}
      />
    )).toMatchSnapshot();
  });

  /**
   * TEST 2: className and activeClassName properties
   * The `className` property should force the component to use a different
   * class name instead of the default.
   */
  it('should render component with custom class names', () => {
    const component = mount(
      <Notification
        message={mockNotification.message}
        activeClassName="foo-active"
        className="foo"
      />
    );

    // Component should have custom classname
    expect(component).toMatchSnapshot();

    component.setProps({ isActive: true });

    // Component should have custom *active* class name
    expect(component).toMatchSnapshot();
  });

  /**
   * TEST 3: onClick property
   * After the action is clicked, the onClick function should be called once.
   */
  it('should call `onClick` function when action is clicked', () => {
    const handleClick = jest.fn();

    const component = mount(
      <Notification
        message={mockNotification.message}
        action={mockNotification.action}
        onClick={handleClick}
      />
    );

    component.find(actionClassName).simulate('click');

    expect(handleClick.mock.calls.length).toBe(1);
  });

  /**
   * TEST 4a: onDismiss property with interval
   * After the time out interval, the onDismiss function should be called once.
   */
  it('should call `onDismiss` function after timeout interval', (done) => {
    const handleDismiss = jest.fn();

    // To prevent errors, set the Jasmine the timeout to a longer interval
    const originalJasmineTimeOutInterval = jasmine.DEFAULT_TIMEOUT_INTERVAL;
    jasmine.DEFAULT_TIMEOUT_INTERVAL = mockNotification.dismissAfter;

    const component = mount(
      <Notification
        message={mockNotification.message}
        dismissAfter={mockNotification.dismissAfter}
        onDismiss={handleDismiss}
      />
    );

    component.setProps({ isActive: true });

    // Before the interval, onDismiss should not be called
    expect(handleDismiss.mock.calls.length).toBe(0);

    // After the interval, expect onDismiss to be called once
    setTimeout(() => {
      expect(handleDismiss.mock.calls.length).toBe(1);
      jasmine.DEFAULT_TIMEOUT_INTERVAL = originalJasmineTimeOutInterval;
      done();
    }, mockNotification.dismissAfter);
  });

  /**
   * TEST 4b: onDismiss property with not interval
   * If `dismissAfter` is falue, the onDismiss function should not be called.
   */
  it('should not call onDismiss when interval is not set', () => {
    const handleDismiss = jest.fn();

    const component = shallow(
      <Notification
        message={mockNotification.message}
        dismissAfter={false}
        onDismiss={handleDismiss}
      />
    );

    component.setProps({ isActive: true });

    expect(handleDismiss.mock.calls.length).toBe(0);
  });

  /**
   * TEST 5a: custom bar styles
   * When custom style properties are provided, those styles should override
   * the default styles.
   */
  it('should render custom bar style', () => {
    const component = shallow(
      <Notification
        message={mockNotification.message}
        barStyle={mockNotification.barStyle}
        activeBarStyle={mockNotification.activeBarStyle}
      />
    );

    // Component should have custom bar style
    chai.expect(component).to.have.style('color', '#000000');
    expect(component).toMatchSnapshot();

    component.setProps({ isActive: true });

    // Component should have custom *active* bar style
    chai.expect(component).to.have.style('color', '#FFFFFF');
    expect(component).toMatchSnapshot();
  });

  /**
   * TEST 5b: custom action styles
   * When custom style properties are provided, those styles should override
   * the default styles.
   *
  it('should render custom action style', () => {
    const component = mount(
      <Notification
        message={mockNotification.message}
        action={mockNotification.action}
        actionStyle={mockNotification.actionStyle}
      />
    );

    // Component should have custom action style
    const action = component.find(actionClassName);
    chai.expect(action).to.have.style('color', '#EEEEEE');
    expect(action).toMatchSnapshot();
  });*/
});
