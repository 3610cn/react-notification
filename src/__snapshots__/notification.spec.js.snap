exports[`Notification should render component with custom class names 1`] = `
<Notification
  activeClassName="foo-active"
  className="foo"
  dismissAfter={2000}
  isActive={false}
  message="Foo">
  <div
    className="notification-bar foo"
    style={
      Object {
        "MozTransform": "translatez(0)",
        "MozTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "OTransform": "translatez(0)",
        "OTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebKitTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebkitTransform": "translatez(0)",
        "background": "#212121",
        "borderRadius": "5px",
        "borderSizing": "border-box",
        "bottom": "2rem",
        "boxShadow": "0 0 1px 1px rgba(10, 10, 11, .125)",
        "color": "#fafafa",
        "cursor": "default",
        "font": "1rem normal Roboto, sans-serif",
        "left": "-100%",
        "margin": 0,
        "msTransform": "translatez(0)",
        "msTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "padding": "1rem",
        "position": "fixed",
        "transform": "translatez(0)",
        "transition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "width": "auto",
      }
    }>
    <div
      className="notification-bar-wrapper">
      <span
        className="notification-bar-message">
        Foo
      </span>
    </div>
  </div>
</Notification>
`;

exports[`Notification should render component with custom class names 2`] = `
<Notification
  activeClassName="foo-active"
  className="foo"
  dismissAfter={2000}
  isActive={true}
  message="Foo">
  <div
    className="notification-bar foo-active foo"
    style={
      Object {
        "MozTransform": "translatez(0)",
        "MozTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "OTransform": "translatez(0)",
        "OTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebKitTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebkitTransform": "translatez(0)",
        "background": "#212121",
        "borderRadius": "5px",
        "borderSizing": "border-box",
        "bottom": "2rem",
        "boxShadow": "0 0 1px 1px rgba(10, 10, 11, .125)",
        "color": "#fafafa",
        "cursor": "default",
        "font": "1rem normal Roboto, sans-serif",
        "left": "1rem",
        "margin": 0,
        "msTransform": "translatez(0)",
        "msTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "padding": "1rem",
        "position": "fixed",
        "transform": "translatez(0)",
        "transition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "width": "auto",
      }
    }>
    <div
      className="notification-bar-wrapper">
      <span
        className="notification-bar-message">
        Foo
      </span>
    </div>
  </div>
</Notification>
`;

exports[`Notification should render component with message type element 1`] = `
<Notification
  activeClassName="notification-bar-active"
  dismissAfter={2000}
  isActive={false}
  message={
    <div
      className="notification-bar-custom-message">
      Foo
    </div>
  }>
  <div
    className="notification-bar"
    style={
      Object {
        "MozTransform": "translatez(0)",
        "MozTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "OTransform": "translatez(0)",
        "OTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebKitTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebkitTransform": "translatez(0)",
        "background": "#212121",
        "borderRadius": "5px",
        "borderSizing": "border-box",
        "bottom": "2rem",
        "boxShadow": "0 0 1px 1px rgba(10, 10, 11, .125)",
        "color": "#fafafa",
        "cursor": "default",
        "font": "1rem normal Roboto, sans-serif",
        "left": "-100%",
        "margin": 0,
        "msTransform": "translatez(0)",
        "msTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "padding": "1rem",
        "position": "fixed",
        "transform": "translatez(0)",
        "transition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "width": "auto",
      }
    }>
    <div
      className="notification-bar-wrapper">
      <span
        className="notification-bar-message">
        <div
          className="notification-bar-custom-message">
          Foo
        </div>
      </span>
    </div>
  </div>
</Notification>
`;

exports[`Notification should render component with message type string 1`] = `
<Notification
  activeClassName="notification-bar-active"
  dismissAfter={2000}
  isActive={false}
  message="Foo">
  <div
    className="notification-bar"
    style={
      Object {
        "MozTransform": "translatez(0)",
        "MozTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "OTransform": "translatez(0)",
        "OTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebKitTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "WebkitTransform": "translatez(0)",
        "background": "#212121",
        "borderRadius": "5px",
        "borderSizing": "border-box",
        "bottom": "2rem",
        "boxShadow": "0 0 1px 1px rgba(10, 10, 11, .125)",
        "color": "#fafafa",
        "cursor": "default",
        "font": "1rem normal Roboto, sans-serif",
        "left": "-100%",
        "margin": 0,
        "msTransform": "translatez(0)",
        "msTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "padding": "1rem",
        "position": "fixed",
        "transform": "translatez(0)",
        "transition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
        "width": "auto",
      }
    }>
    <div
      className="notification-bar-wrapper">
      <span
        className="notification-bar-message">
        Foo
      </span>
    </div>
  </div>
</Notification>
`;

exports[`Notification should render custom bar style 1`] = `
<div
  className="notification-bar"
  style={
    Object {
      "MozTransform": "translatez(0)",
      "MozTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "OTransform": "translatez(0)",
      "OTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "WebKitTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "WebkitTransform": "translatez(0)",
      "background": "#212121",
      "borderRadius": "5px",
      "borderSizing": "border-box",
      "bottom": "2rem",
      "boxShadow": "0 0 1px 1px rgba(10, 10, 11, .125)",
      "color": "#000000",
      "cursor": "default",
      "font": "1rem normal Roboto, sans-serif",
      "left": "-100%",
      "margin": 0,
      "msTransform": "translatez(0)",
      "msTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "padding": "1rem",
      "position": "fixed",
      "transform": "translatez(0)",
      "transition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "width": "auto",
    }
  }>
  <div
    className="notification-bar-wrapper">
    <span
      className="notification-bar-message">
      Foo
    </span>
  </div>
</div>
`;

exports[`Notification should render custom bar style 2`] = `
<div
  className="notification-bar notification-bar-active"
  style={
    Object {
      "MozTransform": "translatez(0)",
      "MozTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "OTransform": "translatez(0)",
      "OTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "WebKitTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "WebkitTransform": "translatez(0)",
      "background": "#212121",
      "borderRadius": "5px",
      "borderSizing": "border-box",
      "bottom": "2rem",
      "boxShadow": "0 0 1px 1px rgba(10, 10, 11, .125)",
      "color": "#FFFFFF",
      "cursor": "default",
      "font": "1rem normal Roboto, sans-serif",
      "left": "1rem",
      "margin": 0,
      "msTransform": "translatez(0)",
      "msTransition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "padding": "1rem",
      "position": "fixed",
      "transform": "translatez(0)",
      "transition": ".5s cubic-bezier(0.89, 0.01, 0.5, 1.1)",
      "width": "auto",
    }
  }>
  <div
    className="notification-bar-wrapper">
    <span
      className="notification-bar-message">
      Foo
    </span>
  </div>
</div>
`;
